keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1362 @@
+   and in issue [kubernetes/kubernetes#114877](https://github.com/kubernetes/kubernetes/issues/114877).
+2. Similar to the first point, it is difficult to use PDBs for applications that can have a variable
+   number of pods; for example applications with a configured horizontal pod autoscaler (HPA). These
+   applications cannot be disrupted during a low load when they have only pod. However, it is
+   possible to disrupt the pods during a high load without experiencing application downtime. If
+   the minimum number of pods is 1, PDBs cannot be used without blocking the node drain. This has
+   been discussed in issue [kubernetes/kubernetes#93476](https://github.com/kubernetes/kubernetes/issues/93476).
+3. Graceful deletion of DaemonSet pods is currently only supported as part of (Linux) graceful node
+   shutdown. The length of the shutdown is again not application specific and is set cluster-wide
+   (optionally by priority) by the cluster admin. This does not take into account
+   `.spec.terminationGracePeriodSeconds` of each pod and may cause premature termination of
+   the application. This has been discussed in issue [kubernetes/kubernetes#75482](https://github.com/kubernetes/kubernetes/issues/75482)
+   and in issue [kubernetes-sigs/cluster-api#6158](https://github.com/kubernetes-sigs/cluster-api/issues/6158).
+4. Descheduler does not allow postponing eviction for applications that are unable to be evicted
+   immediately. This can result in descheduling of incorrect set of pods. This is outlined in the
+   KEP [kubernetes-sigs/descheduler#1354](https://github.com/kubernetes-sigs/descheduler/pull/1354).
+
+### Goals
+
+- Introduce new `evacuation.coordination.k8s.io` API.
+- Introduce evacuation controller.
+
+### Non-Goals
+
+- Synchronizing of the evacuation status to the pod status.
+- Introduce the evacuation concept even for types other than pods.
+
+## Proposal
+
+We will introduce a new term called evacuation. This is a contract between the evacuation instigator,
+the evacuee, and the evacuator. The contract is enforced by the API and an evacuation controller.
+We can think of evacuation as a managed and safer alternative to eviction.
+
+### Evacuation Instigator
+
+The evacuation instigator can be any entity in the system: node maintenance controller, descheduler,
+cluster autoscaler, or any application/controller interfacing with the affected application/pods
+(evacuee).
+
+The instigator's responsibility is to communicate an intent to a pod that it should be evacuated,
+according to the instigator's own internal rules. It should reconcile its intent in case the intent
+is removed by a third party. And it should remove its intent when the evacuation is no longer necessary.
+
+Example evacuation triggers:
+- Node maintenance controller: node maintenance triggered by an admin.
+- Descheduler: descheduling triggered by a descheduling rule.
```

Comment:
If the descheduler requests an eviction, what thing is being evacuated?

(the node maintenance and cluster autoscaler examples are easier: you're evacuating an entire node)

Reply:
A single pod or multiple pods. The descheduler can use it as a new mechanism instead of eviction.

Reply:
OK, so people typically think of “evacuate” as a near synonym of “drain” - you drain a node, you evacuate a rack or zone full of servers. Saying that you can evacuate a Pod might make people think its containers all get stopped, or just confuse readers. We do need to smooth out how we can teach this.

Reply:
It seems it can be used in both scenarios https://www.merriam-webster.com/grammar/can-you-evacuate-people. 
Evacuation of containers doesn't make sense because they are tied to the pod lifecycle. But, I guess it could be confusing if we do not make it explicitly clear what we are targeting.

Reply:
Thing is, Kubernetes users typically  - before we accept this KEP - uses “evacuate” as a synonym for drain.

I'm (still) fine with the API idea, and still concerned about the naming.

Reply:
Just to +1 the potential confusion of the term "evacuation".
Is it okay to have a glossary of terms for "evacuation", "eviction", and "drain" (or any other potentially confusing terms) added somewhere in this KEP?

Reply:
I can include it in the KEP. And yes, we are going to change the name to something more suitable.

Reply:
The API is renamed to EvictionRequest now, see https://github.com/kubernetes/enhancements/pull/4565#discussion_r1867777352 for more details